### DISCUSSION TITLE
Render HTML pages from memory

### DISCUSSION THREAD

**Discussion by anna-hope** [servo/servo#34777] [Q&A]
I am building a proof-of-concept application that includes the ability to read and store [web archives.](https://github.com/ticky/webarchive) I would like to be able to render the HTML and other resources of an archived page from within the application, using an embedded browser.

I thought Servo might be a good option for this, at least at the proof of concept stage, so that I could render web content programmatically. But after a couple of days of reading through the docs and the code (as well as some of the code for Verso) I haven't been able to find a clear way of rendering HTML from a `[u8]` that I have in memory (and not a URL).

I have looked at [`servo::Servo`](https://doc.servo.org/servo/struct.Servo.html), but don't understand the project enough to put the pieces together, and am not even sure that's the right level of abstraction to be looking at, given what I'm trying to do.

I would be grateful if someone with more knowledge of Servo could provide an example, or point me towards an API that would put me on the right track for this!

**Comment by jdm** [comment]
My first thought here is that you could create a custom [protocol handler](https://doc.servo.org/net/protocols/trait.ProtocolHandler.html)—this would allow you to translate a URL like `mem://39c1c3e2-dfe3-4be7-a692-3657b021ca63` to your u8 slice and send the data in the slice through the network stack. The embedding layer is [responsible](https://doc.servo.org/compositing/windowing/trait.EmbedderMethods.html#method.get_protocol_handlers) for [setting up](https://doc.servo.org/net/protocols/struct.ProtocolRegistry.html#method.register) these custom protocol handlers.

**Reply by jdm** [reply]
I don't have any ideas that bypass using a URL—fundamentally, Servo is a web rendering engine that uses URLs, and hard to integrate data into it that can't be mapped to one. For more URL-based alternatives, you could create either a data or srcdoc URL from your in-memory data and load that.

**Reply by jdm** [reply]
servo::Servo is sort of the right level of abstraction, with the caveat that we don't have great examples for what it takes to just render a particular piece of content and do very little else. https://github.com/jdm/servo-embedding-example/blob/main/src/main_glutin.rs is probably the best reference for that right now.

**Reply by anna-hope** [reply]
> I don't have any ideas that bypass using a URL—fundamentally, Servo is a web rendering engine that uses URLs, and hard to integrate data into it that can't be mapped to one. For more URL-based alternatives, you could create either a data or srcdoc URL from your in-memory data and load that.

I appreciate the quick response! 

`data` URLs (maybe combined with `srcdoc`) could be the way to go. 

MDN's docs say that many browsers limit the length of data URLs they accept: https://developer.mozilla.org/en-US/docs/Web/URI/Schemes/data#length_limitations. The limits for most browsers they list are high enough that it shouldn't cause problems for me. Do you know if Servo currently limits the length of data URLs?  


> servo::Servo is sort of the right level of abstraction, with the caveat that we don't have great examples for what it takes to just render a particular piece of content and do very little else. https://github.com/jdm/servo-embedding-example/blob/main/src/main_glutin.rs is probably the best reference for that right now.

Thank you for linking the example! I'll look through it tomorrow.

**Reply by jdm** [reply]
I don't see a limit in Servo's implementation during a quick skim of https://github.com/servo/rust-url/tree/main/data-url/src and https://github.com/servo/servo/blob/adc814d7ff79a0e908317727904ba361acbf209a/components/net/protocols/data.rs, but I might be overlooking it.
